File: Vandame/matlab_code/Cfind_shape.m
threshold = -144 ;%-80; %(normalize dB)
threshold = 10^(0.05*threshold);
global Data
NB_pixel_min = 50; % perfect = [1] %Warning
cmerge_threshold = 5;
infomin = 70;
canvas = 10*[200 1];
se = strel('disk',3);
% cmerge_threshold = 1.3;
disp('Find shape');
tab_NB_object = zeros(NB_folder,NB_file);
[dim1,dim2] = size(Data.spectroinfo.gridT);
%%
total = zeros(NB_folder,1);
for i = 1 : NB_folder
    Data.directory(i).pointer=zeros(NB_file,dim1,dim2);
    for j = 1 : NB_file
        %% display time frequency
        
        pointer =  (squeeze(Data.directory(i).spectro(j,:,:))) > threshold;
        pointer = imclose(pointer,se);
        Data.directory(i).pointer(j,:,:) = pointer;
        canvas = diag(canvas);
        
        %% isolation shape
        obj = bwconncomp(pointer);
        for zz = 1 : obj.NumObjects
             total(i) = total(i)+length(obj.PixelIdxList{zz});
        end
        center =regionprops(obj,'Centroid');
        center = cat(1,center.Centroid);
        center = fliplr(center)*canvas;
        link = linkage(center,'ward','euclidean');
        
        
        
        clust = cluster(link,'cutoff',cmerge_threshold,'criterion','distance');
        %clust = clusterdata(center,cmerge_threshold);
        nobj = cell(1,max(clust));
        obj.NumObjects = max(clust);
        
        for m = 1 : length(clust)
            nobj{clust(m)} = cat(1,nobj{clust(m)},obj.PixelIdxList{m});
        end
        
        obj.PixelIdxList = nobj;
        clear nobj locs pks link clusr center delta;
        %% delete little object
        NB_object = obj.NumObjects;
        for k = 1 : obj.NumObjects
            [NB_pixel_shape,~]=size(obj.PixelIdxList{k});
            if NB_pixel_shape<NB_pixel_min
                obj.PixelIdxList{k}=[];
                NB_object=NB_object-1;
            end
        end
        obj.NumObjects = NB_object;
        obj.PixelIdxList = obj.PixelIdxList(~cellfun('isempty', obj.PixelIdxList));
        
        Data.directory(i).obj(j).Connectivity=obj.Connectivity;
        Data.directory(i).obj(j).ImageSize=obj.ImageSize;
        Data.directory(i).obj(j).NumObjects=obj.NumObjects;
        Data.directory(i).obj(j).PixelIdxList=obj.PixelIdxList;
        tab_NB_object(i,j) = NB_object;
        
    end
end

disp(tab_NB_object);
disp(['sum : ',num2str(sum(tab_NB_object(:)))]);
save('total.mat','total');
%%
close all;
clear ans threshold canvas pointer canvas_min cmerge_threshold m clust 
clear center NB_pixel_shape i j k NB_pixel_min obj NB_object subp fig se 
clear dim1 dim2 fig

File: Vandame/ChatGPT_MATtoPY/Cfind_shape.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.ndimage import label
from scipy.cluster.hierarchy import linkage, fcluster
from scipy.spatial.distance import pdist
from skimage.morphology import closing, disk

threshold = -144
threshold = 10 ** (0.05 * threshold)

NB_pixel_min = 50
cmerge_threshold = 5
infomin = 70
canvas = 10 * np.array([200, 1])
se = disk(3)

print('Find shape')

tab_NB_object = np.zeros((NB_folder, NB_file))
dim1, dim2 = Data.spectroinfo.gridT.shape

total = np.zeros(NB_folder)

for i in range(NB_folder):
    Data.directory[i].pointer = np.zeros((NB_file, dim1, dim2))
    for j in range(NB_file):
        pointer = Data.directory[i].spectro[j] > threshold
        pointer = closing(pointer, selem=se)
        Data.directory[i].pointer[j] = pointer.astype(int)
        canvas = np.diag(canvas)

        obj, num_objects = label(pointer)
        for zz in range(1, num_objects + 1):
            total[i] += np.sum(obj == zz)

        center = np.array([prop.centroid for prop in regionprops(obj)])
        center = np.fliplr(center) * canvas
        link = linkage(center, method='ward', metric='euclidean')

        clust = fcluster(link, cmerge_threshold, criterion='distance')
        nobj = [np.array([]) for _ in range(clust.max())]
        obj.NumObjects = clust.max()

        for m in range(len(clust)):
            nobj[clust[m] - 1] = np.concatenate((nobj[clust[m] - 1], obj[obj == m + 1]))

        obj.PixelIdxList = [n for n in nobj if len(n) > 0]
        obj.NumObjects = len(obj.PixelIdxList)
        Data.directory[i].obj[j].Connectivity = obj.Connectivity
        Data.directory[i].obj[j].ImageSize = obj.ImageSize
        Data.directory[i].obj[j].NumObjects = obj.NumObjects
        Data.directory[i].obj[j].PixelIdxList = obj.PixelIdxList
        tab_NB_object[i, j] = obj.NumObjects

print(tab_NB_object)
print('sum:', np.sum(tab_NB_object))

total = total.astype(int)
np.save('total.npy', total)
